data/adapters/helpers.py

```diff
@@ -416,11 +418,112 @@ def get_ref_seq_by_spdi(spdi, species='human'):
+        By default: validate ref allele for both SNVs and indels, and skip those failed validation variants instead of correcting the ref allele for them automatically.
+    '''
+    variant_json = {}
+    skipped_message = None
+    format = None
+    spdi = None
+
+    if len(variant_id.split(':')) == 4:
+        format = 'spdi'
+        chr_spdi = variant_id.split(':')[0]
+        chr, pos_start, ref, alt = split_spdi(variant_id)
+    elif len(variant_id.split('-')) == 4:
+        format = 'vcf'
+        chr, pos_start, ref, alt = variant_id.split('-')
+    else:
+        skipped_message = {'variant_id': variant_id,
+                           'reason': 'Unable to parse this variant id'}
+        return variant_json, skipped_message
+
+    # Note: we convert the position to 1-based for spdi format id here, and input format as 'gnomad' when calling translator from ga4gh.vrs, since translate_from spdi doesn't include validation step currently
+    # Add special case when ref or alt is empty - they are not accepted in gnomad/vcf format
+    if format == 'spdi':
+        if ref == '':
+            # no need to validate ref allele
+            spdi = f'{chr_spdi}:{pos_start}:{ref}:{alt}'
+        elif alt == '':
+            ref_genome = get_ref_seq_by_spdi(variant_id)
+            if ref != ref_genome:
+                skipped_message = {'variant_id': variant_id,
+                                   'reason': 'Ref allele mismatch'}
+                return variant_json, skipped_message
+            spdi = f'{chr_spdi}:{pos_start}:{ref}:{alt}'
+
+    if spdi is None:
+        # do validation and normalization for both single nucleotide variants and multiple nucleotide variants, with translator from ga4gh.vrs
+        # though SNV doesn't need the normalization part
+        if format == 'spdi':
+            pos_start = pos_start + 1
+        seq_repo = get_seqrepo('human')
+        data_proxy = SeqRepoDataProxy(seq_repo)
+        translator = AlleleTranslator(data_proxy)
+        try:
+            spdi = build_spdi(chr, pos_start, ref,
+                              alt, translator, seq_repo, assembly, validate_SNV, correct_ref_allele)
+        except ValueError as e:
+            skipped_message = {'variant_id': variant_id,
+                               'reason': 'Ref allele mismatch'}
+            return variant_json, skipped_message
+    if len(spdi) < 254:
+        _id = spdi
+    else:
+        allele = build_allele(chr, pos_start, ref,
+                              alt, translator, seq_repo, assembly)
+        _id = allele.digest
+
+    variation_type = 'SNP'  # should be SNV more broadly
+    if len(ref) < len(alt):
+        variation_type = 'insertion'
+    elif len(ref) > len(alt):
+        variation_type = 'deletion'
+    elif len(ref) > 1:
+        # e.g. NC_000018.10:31546003:AA:TG
+        variation_type = 'deletion-insertion'
```

Comment:
What about 
elif len(ref) == len(alt) == 1:
    variation_type = 'substitution'

Reply:
so replace SNP with substitution? I wanted to keep it the same as those we already loaded from favor for now

Reply:
Ok that's fine, I forgot there was a default value